File: daily/Sliding Window/76-Minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ""
        
        ans = float('inf'), None, None
        left = 0

        formed = 0
        frequency_t = Counter(t)
        required = len(frequency_t)
        current_window = defaultdict(int)

        for right in range(len(s)):
            char_r = s[right]
            current_window[char_r] += 1

            if char_r in frequency_t and current_window[char_r] == frequency_t[char_r]:
                formed += 1
            
            while left <= right and formed == required:
                char_l = s[left]

                if right - left + 1 < ans[0]:
                    ans = right - left + 1, left, right
                current_window[char_l] -= 1
                if char_l in frequency_t and current_window[char_l] < frequency_t[char_l]:
                    formed -= 1
                left += 1
        return "" if ans[0] == float('inf') else s[ans[1] : ans[2] + 1]

File: daily/Sliding Window/76-Minimum-window-substring.go
package daily

import "math"

func minWindow(s string, t string) string {
	if len(s) == 0 || len(t) == 0 {
		return ""
	}

	ans := []int{math.MaxInt64, 0, 0}
	left := 0

	formed := 0
	frequency_t := make(map[byte]int)
	for i := 0; i < len(t); i++ {
		frequency_t[t[i]]++
	}
	required := len(frequency_t)
	current_window := make(map[byte]int)

	for right := range s {
		char_r := s[right]
		current_window[char_r]++

		if count_t, ok := frequency_t[char_r]; ok && current_window[char_r] == count_t {
			formed++
		}

		for left <= right && formed == required {
			char_l := s[left]

			if right-left+1 < ans[0] {
				ans = []int{right - left + 1, left, right}
			}
			current_window[char_l]--
			if count_t, ok := frequency_t[char_l]; ok && current_window[char_l] < count_t {
				formed--
			}
			left++
		}
	}
	if ans[0] == math.MaxInt64 {
		return ""
	}
	return s[ans[1] : ans[2]+1]
}
